File: src/main/java/org/kite9/diagram/visualization/planarization/rhd/links/ContradictionHandler.java
package org.kite9.diagram.visualization.planarization.rhd.links;

import org.kite9.diagram.common.BiDirectional;
import org.kite9.diagram.model.Connected;
import org.kite9.diagram.model.Connection;
import org.kite9.diagram.model.Container;
import org.kite9.diagram.model.position.Direction;
import org.kite9.diagram.model.position.Layout;
import org.kite9.diagram.visualization.planarization.rhd.links.LinkManager.LinkDetail;

/**
 * Contains code for managing/checking for contradictions in links.
 * 
 * @author robmoffat
 *
 */
public interface ContradictionHandler {

	public abstract void checkOrdinalContradiction(Layout l, Direction d,
			Connected from, Connected to, Container fromC, Connection c);

	public abstract void checkForContainerContradiction(Connection c);

	public abstract Direction checkContradiction(Direction ad, boolean aOrdering,
			int aRank, Iterable<BiDirectional<Connected>> ac, Direction bd, boolean bOrdering, int bRank,
			Iterable<BiDirectional<Connected>> bc, Layout containerLayout);

	public abstract Direction checkContradiction(LinkDetail ld1, LinkDetail ld2,
			Layout containerLayout);

	public abstract void setContradiction(BiDirectional<Connected> bic);

	public abstract void setContradicting(Iterable<BiDirectional<Connected>> connections);

}

File: src/main/kotlin/org/kite9/diagram/visualization/planarization/rhd/links/ContradictionHandler.kt
package org.kite9.diagram.visualization.planarization.rhd.links

import org.kite9.diagram.common.BiDirectional
import org.kite9.diagram.model.Connected
import org.kite9.diagram.model.Connection
import org.kite9.diagram.model.position.Direction
import org.kite9.diagram.model.position.Layout
import org.kite9.diagram.visualization.planarization.rhd.links.LinkManager.LinkDetail

/**
 * Contains code for managing/checking for contradictions in links.
 *
 * @author robmoffat
 */
interface ContradictionHandler {

    fun checkForContainerContradiction(c: Connection)

    fun checkContradiction(
        ad: Direction, aOrdering: Boolean,
        aRank: Int, ac: Iterable<BiDirectional<Connected>>, bd: Direction, bOrdering: Boolean, bRank: Int,
        bc: Iterable<BiDirectional<Connected>>, containerLayout: Layout
    ): Direction

    fun checkContradiction(
        ld1: LinkDetail, ld2: LinkDetail,
        containerLayout: Layout
    ): Direction

    fun setContradiction(bic: BiDirectional<Connected>, dontRender: Boolean)
    fun setContradicting(connections: Iterable<BiDirectional<Connected>>, dontRender: Boolean)
}